grpc/noncebalancer/noncebalancer.go

```diff
@@ -0,0 +1,116 @@
+package noncebalancer
+
+import (
+	"errors"
+
+	"github.com/letsencrypt/boulder/nonce"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+)
+
+const (
+	// Name is the name used to register the nonce balancer with the gRPC
+	// runtime.
+	Name = "nonce"
+
+	// SRVResolverScheme is the scheme used to invoke an instance of the SRV
+	// resolver which will use the noncebalancer to pick backends. It would be
+	// ideal to export this from the SRV resolver package but that package is
+	// internal.
+	SRVResolverScheme = "nonce-srv"
+)
+
+var errMissingPrefixCtxKey = errors.New("nonce.PrefixCtxKey value required in RPC context")
+var errMissingHMACKeyCtxKey = errors.New("nonce.HMACKeyCtxKey value required in RPC context")
+var errInvalidPrefixCtxKeyType = errors.New("nonce.PrefixCtxKey value in RPC context must be a string")
+var errInvalidHMACKeyCtxKeyType = errors.New("nonce.HMACKeyCtxKey value in RPC context must be a string")
+
+// Balancer implements the base.PickerBuilder interface. It's used to factory
+// new balancer.Picker instances. It should only be used by nonce-service
+// clients.
+type Balancer struct{}
+
+// Compile-time assertion that *Balancer implements the base.PickerBuilder
+// interface.
+var _ base.PickerBuilder = (*Balancer)(nil)
+
+// Build implements the base.PickerBuilder interface. It is called by the gRPC
+// runtime when the balancer is first initialized and when the set of backend
+// (SubConn) addresses changes. It is responsible for initializing the Picker's
+// backends map and returning a balancer.Picker.
+func (b *Balancer) Build(buildInfo base.PickerBuildInfo) balancer.Picker {
+	return &Picker{
+		backends: buildInfo.ReadySCs,
+	}
+}
+
+// Picker implements the balancer.Picker interface. It picks a backend (SubConn)
+// based on the nonce prefix contained in each request's Context.
+type Picker struct {
+	backends        map[balancer.SubConn]base.SubConnInfo
+	prefixToBackend map[string]balancer.SubConn
+}
+
+// Compile-time assertion that *Picker implements the balancer.Picker interface.
+var _ balancer.Picker = (*Picker)(nil)
+
+// Pick implements the balancer.Picker interface. It is called by the gRPC
+// runtime for each RPC message. It is responsible for picking a backend
+// (SubConn) based on the context of each RPC message.
+func (p *Picker) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
+	var result balancer.PickResult
```

Comment:
Definitely a nit, but I'd prefer if we didn't pre-declare this here, and instead said `return blancer.PickResult{}, someError` in all of the early-return paths.

As it currently stands, a reader gets to line 107 and thinks "oh no, when have other fields on this result been set?" and has to re-read the function to verify that `result` hasn't been modified in the mean time.